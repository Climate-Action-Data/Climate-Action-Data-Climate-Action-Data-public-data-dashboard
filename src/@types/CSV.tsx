export enum CSVExportTypes {
  PROJECT = `project`,
  UNIT = `unit`,
}

export enum CSVExportFilenames {
  PROJECT_SEARCH = `project-search-export`,
  UNIT_SEARCH = `unit-search-export`,
  PROJECT_COMPARE = `project-compare-export`,
}

export type ExportFilters = {
  keywords: string
  count: number
  offset: number
  sortBy: string
  direction: string
}

export type ExportProjectFilters = {
  standards: string[]
  methodologies: string[]
  sectors: string[]
  countries: string[]
  creditingPeriodStart?: string
  creditingPeriodEnd?: string
}

export type ExportUnitFilters = {
  status: string[]
  standards: string[]
  methodologies: string[]
  sectors: string[]
  countries: string[]
  minYear?: number
  maxYear?: number
}

export type ExportCompleteFilters = ExportFilters & (ExportUnitFilters | ExportProjectFilters)
