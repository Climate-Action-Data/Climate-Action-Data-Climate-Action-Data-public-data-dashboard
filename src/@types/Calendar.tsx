import { format, setDay, setMonth } from 'date-fns'

export const MONTHS_IN_YEAR = 12
export const DAYS_IN_WEEK = 7

export const getMonthNames = () => {
  const monthNames = []
  for (let i = 0; i < MONTHS_IN_YEAR; i++) {
    monthNames.push(format(setMonth(new Date(), i), `MMM`))
  }
  return monthNames
}

export const getWeekdayNames = () => {
  const weekdayNames = []
  for (let i = 0; i < DAYS_IN_WEEK; i++) {
    weekdayNames.push(format(setDay(new Date(), i), `EEEEEE`))
  }
  return weekdayNames
}
