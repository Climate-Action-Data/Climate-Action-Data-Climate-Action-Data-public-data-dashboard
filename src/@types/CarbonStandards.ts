export enum CarbonStandards {
    VCS = "vcs",
    GCC = "gcc",
    ECO = "eco"
}