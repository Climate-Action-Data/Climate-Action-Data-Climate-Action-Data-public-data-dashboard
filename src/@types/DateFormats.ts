export enum DateFormats {
  YYYY_MM_DD = `yyyy/MM/dd`,
  YYYY_MM_DD_HH_MM = `yyyy/MM/dd HH:mm`,
  YYYY_MM_DD_HH_MM_SS = `yyyy/MM/dd HH:mm:ss`,
}
