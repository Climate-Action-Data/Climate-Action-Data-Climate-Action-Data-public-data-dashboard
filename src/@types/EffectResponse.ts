export interface EffectResponse<T> {
  data?: T
  error?: ErrorResponse
}

export interface ErrorResponse {
  message: string
  code: string
}
