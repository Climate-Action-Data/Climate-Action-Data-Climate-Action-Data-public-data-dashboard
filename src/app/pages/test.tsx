'use client'
import { type NextPage } from "next";

const Test: NextPage = () => {
    return (
        <div>
            <h1>Hello</h1>
            <p>data</p>
        </div>
    )
}

export default Test