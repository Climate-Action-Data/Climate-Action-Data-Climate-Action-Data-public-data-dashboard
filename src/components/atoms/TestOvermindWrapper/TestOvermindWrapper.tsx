import { createOvermind } from 'overmind'
import { Provider } from 'overmind-react'
import { config } from '@/overmind'
import { PropsWithChildren } from 'react'

const overmind = createOvermind(config)

export const TestOvermindWrapper = (props: PropsWithChildren) => {
  return <Provider value={overmind}>{props.children}</Provider>
}
