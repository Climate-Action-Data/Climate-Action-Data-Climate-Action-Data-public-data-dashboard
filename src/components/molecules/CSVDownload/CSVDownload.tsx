import { CSVExportTypes } from '@/@types/CSV'
import { ProjectSearchFilterValues, UnitSearchFilterValues } from '@/@types/ProjectSearchFilterValues'
import { ToastVariants } from '@/@types/Toast'
import { DownloadIcon } from '@/components/atoms/DownloadIcon/DownloadIcon'
import { useToastHook } from '@/components/atoms/Toast/Toast'
import { useActions, useAppState, useEffects } from '@/overmind'
import { Button, Hide, Text } from '@chakra-ui/react'
import { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'

interface CSVDownloadProps {
  exportType: CSVExportTypes
}

const DEFAULT_BLOB_TYPE = `application/octet-stream`
const DEFAULT_LINK_ELEMENT = `a`
const DEFAULT_DL_ATTRIBUTE = `download`
const DEFAULT_FILE_NAME = `export.csv`
const DEFAULT_LINK_DISPLAY = `none`
const createAndDownload = (data: Blob) => {
  const blob = new Blob([data], { type: DEFAULT_BLOB_TYPE })
  const url = window.URL.createObjectURL(blob)
  const link = document.createElement(DEFAULT_LINK_ELEMENT)
  link.style.display = DEFAULT_LINK_DISPLAY
  link.href = url
  link.setAttribute(DEFAULT_DL_ATTRIBUTE, DEFAULT_FILE_NAME)
  document.body.appendChild(link)
  link.click()
  URL.revokeObjectURL(url)
}

export const CSVDownload = (props: CSVDownloadProps) => {
  const { exportType } = props
  const [preparing, setPreparing] = useState<boolean>(false)
  const { t } = useTranslation(`search`)
  const { setDownloadStatus } = useActions().exports
  const { keywordSearch, selectedProjectSearchFilterValues, selectedUnitSearchFilterValues } = useAppState().searchFilters
  const { hasDownloaded } = useAppState().exports
  const { exportToCSV } = useEffects().exports
  const [newToast] = useToastHook()

  const handleOnClick = async () => {
    setPreparing(true)
    setDownloadStatus(false)
    const downloadStatus = false
    let searchFilters: ProjectSearchFilterValues | UnitSearchFilterValues = selectedProjectSearchFilterValues.searchFilterValues
    if (exportType === CSVExportTypes.UNIT) {
      searchFilters = selectedUnitSearchFilterValues.searchFilterValues
    }
    exportToCSV(exportType, keywordSearch, searchFilters).then((exportData) => {
      if (exportData.data) {
        // Create a blob from the response data
        createAndDownload(exportData.data)
        newToast({ variant: ToastVariants.SUCCESS, message: t(`exportValid`), icon: <DownloadIcon /> })
      } else {
        newToast({ variant: ToastVariants.FAILURE, message: t(`exportFail`), icon: <DownloadIcon /> })
      }
      setPreparing(false)
    })
    setDownloadStatus(downloadStatus)
  }

  useEffect(() => {
    if (hasDownloaded) {
      setPreparing(false)
    }
  }, [hasDownloaded])

  if (preparing) {
    return <Text>{t(`preparing`)}</Text>
  } else {
    return (
      <Button data-testid="export-data" onClick={handleOnClick} variant="hoverOnly" display="flex" gap="4px" fontWeight="500px">
        <Hide below="md">{t(`export`)}</Hide>
        <DownloadIcon />
      </Button>
    )
  }
}
