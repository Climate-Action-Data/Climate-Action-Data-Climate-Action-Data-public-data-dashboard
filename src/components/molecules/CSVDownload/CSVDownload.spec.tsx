import { render, screen, fireEvent } from '@testing-library/react'
import { CSVDownload, createAndDownload } from './CSVDownload'
import { TestOvermindWrapper } from '@/test/TestOvermindWrapper'
import { CSVExportTypes } from '@/@types/CSV'

it(`renders correctly`, () => {
  const { container } = render(
    <TestOvermindWrapper>
      <CSVDownload exportType={CSVExportTypes.PROJECT} />
    </TestOvermindWrapper>,
  )
  expect(container).toMatchSnapshot()
})

it(`renders correctly as unit`, () => {
  const { container } = render(
    <TestOvermindWrapper>
      <CSVDownload exportType={CSVExportTypes.UNIT} />
    </TestOvermindWrapper>,
  )
  expect(container).toMatchSnapshot()
})

it(`renders correctly as unit and click`, () => {
  const { container } = render(
    <TestOvermindWrapper>
      <CSVDownload exportType={CSVExportTypes.UNIT} />
    </TestOvermindWrapper>,
  )
  const downloadButton = screen.getByTestId(`export-data`)
  fireEvent.click(downloadButton)
  expect(container).toMatchSnapshot()
})
it(`renders correctly as pattern and click`, () => {
  const { container } = render(
    <TestOvermindWrapper>
      <CSVDownload exportType={CSVExportTypes.PROJECT} />
    </TestOvermindWrapper>,
  )
  const downloadButton = screen.getByTestId(`export-data`)
  fireEvent.click(downloadButton)
  expect(container).toMatchSnapshot()
})

// Create a test for createAndDownload function in CSVDownload.tsx
//window.URL.createObjectUR
const createUrlMock = jest.fn()
const revokeUrlMock = jest.fn()
beforeAll(() => {
  window.URL.createObjectURL = createUrlMock
  window.URL.revokeObjectURL = revokeUrlMock
})
it(`creates a link and downloads`, () => {
  const testData = new Blob([`test,test2`], { type: `text/csv` })
  createAndDownload(testData)
  const invisibleLink = screen.getByTestId(`invisible-link`)
  expect(createUrlMock).toHaveBeenCalled()
  expect(revokeUrlMock).toHaveBeenCalled()
  expect(invisibleLink).toBeDefined()
})

afterAll(() => {
  jest.clearAllMocks()
})
