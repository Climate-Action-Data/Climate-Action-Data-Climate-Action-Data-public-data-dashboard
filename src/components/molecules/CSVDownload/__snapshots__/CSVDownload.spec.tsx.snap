// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with no pattern 1`] = `
<div>
  <button
    class="chakra-button css-hli6hn"
    data-testid="export-data"
    type="button"
  >
    Export
    <svg
      class="chakra-icon css-n059si"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 15.3255C11.8795 15.3255 11.7673 15.3063 11.6634 15.2679C11.5596 15.2294 11.4609 15.1634 11.3673 15.0698L8.2577 11.9602C8.11925 11.8217 8.04842 11.6477 8.04522 11.4381C8.04201 11.2285 8.11283 11.0513 8.2577 10.9064C8.40257 10.7615 8.58077 10.6865 8.7923 10.6814C9.00383 10.6763 9.18204 10.7461 9.32692 10.891L11.25 12.8141V5.16404C11.25 4.95122 11.3218 4.77302 11.4654 4.62944C11.609 4.48585 11.7872 4.41406 12 4.41406C12.2128 4.41406 12.391 4.48585 12.5346 4.62944C12.6782 4.77302 12.7499 4.95122 12.7499 5.16404V12.8141L14.673 10.891C14.8115 10.7525 14.9881 10.6843 15.2028 10.6862C15.4176 10.6881 15.5974 10.7615 15.7422 10.9064C15.8871 11.0513 15.9595 11.2269 15.9595 11.4333C15.9595 11.6397 15.8871 11.8153 15.7422 11.9602L12.6327 15.0698C12.5391 15.1634 12.4403 15.2294 12.3365 15.2679C12.2327 15.3063 12.1205 15.3255 12 15.3255ZM6.3077 19.5871C5.80257 19.5871 5.375 19.4121 5.025 19.0621C4.675 18.7121 4.5 18.2845 4.5 17.7794V15.8371C4.5 15.6243 4.5718 15.4461 4.7154 15.3025C4.85898 15.1589 5.03718 15.0871 5.25 15.0871C5.46282 15.0871 5.64102 15.1589 5.7846 15.3025C5.92818 15.4461 5.99997 15.6243 5.99997 15.8371V17.7794C5.99997 17.8563 6.03202 17.9268 6.09612 17.991C6.16024 18.0551 6.23077 18.0871 6.3077 18.0871H17.6922C17.7692 18.0871 17.8397 18.0551 17.9038 17.991C17.9679 17.9268 18 17.8563 18 17.7794V15.8371C18 15.6243 18.0718 15.4461 18.2154 15.3025C18.3589 15.1589 18.5371 15.0871 18.75 15.0871C18.9628 15.0871 19.141 15.1589 19.2845 15.3025C19.4281 15.4461 19.5 15.6243 19.5 15.8371V17.7794C19.5 18.2845 19.325 18.7121 18.975 19.0621C18.625 19.4121 18.1974 19.5871 17.6922 19.5871H6.3077Z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`renders correctly with pattern 1`] = `
<div>
  <button
    class="chakra-button css-hli6hn"
    data-testid="export-data"
    type="button"
  >
    Export
    <svg
      class="chakra-icon css-n059si"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 15.3255C11.8795 15.3255 11.7673 15.3063 11.6634 15.2679C11.5596 15.2294 11.4609 15.1634 11.3673 15.0698L8.2577 11.9602C8.11925 11.8217 8.04842 11.6477 8.04522 11.4381C8.04201 11.2285 8.11283 11.0513 8.2577 10.9064C8.40257 10.7615 8.58077 10.6865 8.7923 10.6814C9.00383 10.6763 9.18204 10.7461 9.32692 10.891L11.25 12.8141V5.16404C11.25 4.95122 11.3218 4.77302 11.4654 4.62944C11.609 4.48585 11.7872 4.41406 12 4.41406C12.2128 4.41406 12.391 4.48585 12.5346 4.62944C12.6782 4.77302 12.7499 4.95122 12.7499 5.16404V12.8141L14.673 10.891C14.8115 10.7525 14.9881 10.6843 15.2028 10.6862C15.4176 10.6881 15.5974 10.7615 15.7422 10.9064C15.8871 11.0513 15.9595 11.2269 15.9595 11.4333C15.9595 11.6397 15.8871 11.8153 15.7422 11.9602L12.6327 15.0698C12.5391 15.1634 12.4403 15.2294 12.3365 15.2679C12.2327 15.3063 12.1205 15.3255 12 15.3255ZM6.3077 19.5871C5.80257 19.5871 5.375 19.4121 5.025 19.0621C4.675 18.7121 4.5 18.2845 4.5 17.7794V15.8371C4.5 15.6243 4.5718 15.4461 4.7154 15.3025C4.85898 15.1589 5.03718 15.0871 5.25 15.0871C5.46282 15.0871 5.64102 15.1589 5.7846 15.3025C5.92818 15.4461 5.99997 15.6243 5.99997 15.8371V17.7794C5.99997 17.8563 6.03202 17.9268 6.09612 17.991C6.16024 18.0551 6.23077 18.0871 6.3077 18.0871H17.6922C17.7692 18.0871 17.8397 18.0551 17.9038 17.991C17.9679 17.9268 18 17.8563 18 17.7794V15.8371C18 15.6243 18.0718 15.4461 18.2154 15.3025C18.3589 15.1589 18.5371 15.0871 18.75 15.0871C18.9628 15.0871 19.141 15.1589 19.2845 15.3025C19.4281 15.4461 19.5 15.6243 19.5 15.8371V17.7794C19.5 18.2845 19.325 18.7121 18.975 19.0621C18.625 19.4121 18.1974 19.5871 17.6922 19.5871H6.3077Z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`renders correctly with pattern as project and click 1`] = `
<div>
  <p
    class="chakra-text css-0"
  >
    Preparing download...
  </p>
</div>
`;

exports[`renders correctly with pattern no project 1`] = `
<div>
  <button
    class="chakra-button css-hli6hn"
    data-testid="export-data"
    type="button"
  >
    Export
    <svg
      class="chakra-icon css-n059si"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 15.3255C11.8795 15.3255 11.7673 15.3063 11.6634 15.2679C11.5596 15.2294 11.4609 15.1634 11.3673 15.0698L8.2577 11.9602C8.11925 11.8217 8.04842 11.6477 8.04522 11.4381C8.04201 11.2285 8.11283 11.0513 8.2577 10.9064C8.40257 10.7615 8.58077 10.6865 8.7923 10.6814C9.00383 10.6763 9.18204 10.7461 9.32692 10.891L11.25 12.8141V5.16404C11.25 4.95122 11.3218 4.77302 11.4654 4.62944C11.609 4.48585 11.7872 4.41406 12 4.41406C12.2128 4.41406 12.391 4.48585 12.5346 4.62944C12.6782 4.77302 12.7499 4.95122 12.7499 5.16404V12.8141L14.673 10.891C14.8115 10.7525 14.9881 10.6843 15.2028 10.6862C15.4176 10.6881 15.5974 10.7615 15.7422 10.9064C15.8871 11.0513 15.9595 11.2269 15.9595 11.4333C15.9595 11.6397 15.8871 11.8153 15.7422 11.9602L12.6327 15.0698C12.5391 15.1634 12.4403 15.2294 12.3365 15.2679C12.2327 15.3063 12.1205 15.3255 12 15.3255ZM6.3077 19.5871C5.80257 19.5871 5.375 19.4121 5.025 19.0621C4.675 18.7121 4.5 18.2845 4.5 17.7794V15.8371C4.5 15.6243 4.5718 15.4461 4.7154 15.3025C4.85898 15.1589 5.03718 15.0871 5.25 15.0871C5.46282 15.0871 5.64102 15.1589 5.7846 15.3025C5.92818 15.4461 5.99997 15.6243 5.99997 15.8371V17.7794C5.99997 17.8563 6.03202 17.9268 6.09612 17.991C6.16024 18.0551 6.23077 18.0871 6.3077 18.0871H17.6922C17.7692 18.0871 17.8397 18.0551 17.9038 17.991C17.9679 17.9268 18 17.8563 18 17.7794V15.8371C18 15.6243 18.0718 15.4461 18.2154 15.3025C18.3589 15.1589 18.5371 15.0871 18.75 15.0871C18.9628 15.0871 19.141 15.1589 19.2845 15.3025C19.4281 15.4461 19.5 15.6243 19.5 15.8371V17.7794C19.5 18.2845 19.325 18.7121 18.975 19.0621C18.625 19.4121 18.1974 19.5871 17.6922 19.5871H6.3077Z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`renders correctly with pattern no project and click 1`] = `
<div>
  <p
    class="chakra-text css-0"
  >
    Preparing download...
  </p>
</div>
`;
